### PR TITLE
Fix #11101 - The value can not be NULL. Parameter name: entityType

### LINKED ISSUE #11101: The value can not be NULL. Parameter name: entityType
I am working on a project where i am doing reverse engineering of my database to auto generate entities and DbContexts using **EF Core Power Tools version 1.0.525** in visual studio 2017 Version 15.5, Microsoft.EntityFrameworkCore 2.0.1, Microsoft.EntityFrameworkCore.SqlServer  . 
The reverse engineering is successful, pretty clean auto generated code thumbs to **ErikEJ**. But when am using Linq to Sql queries that involve **inner joins** and **left outer join**. I always run into this exception  **The value can not be NULL. Parameter name: entityType**

below is a sample piece of my code as cannot upload the whole project because of a company regulation and also it being a big. 

``` c#
var query = from bs in this.mDbContext.tblboxs
 join ld in this.pdbContext.tblDicts on bs.BID equals ld.ID 
join mbs in this.crmDbContext.tblStatuses on bs.ID equals mbs.Id into mbsGrps 
from mbs in mbsGrps.DefaultIfEmpty()
 select new XRecord {
     BsID_Archiv = bs.ID,
   };
 return await query.ToListAsync();
```

Remember all the EF-fluent Api mappings and configurations are done for me when i use ** EF Core Power Tools ** and it really pretty clean auto generated code.
 More information
- my tables don't have any relationship constraint(s)  but the relationship is in the data stored in those tables and also i happen to have alots of nullable fields as the project is just used for loading data then later i do SSIS ETL operations.

### PR DESCRIPTION
**Pull request by anpete**
- Validate that a query only uses a single EF IQueryProvider instance.

### CODE REVIEW

File: src/EFCore.Relational/Query/Expressions/SelectExpression.cs [hunk 1]
```diff
@@ -80,6 +81,12 @@ public class SelectExpression : TableExpressionBase
             Alias = queryCompilationContext.CreateUniqueTableAlias(alias);
         }
 
+        internal void DetachContext()
+        {
+            // TODO: Remove dependency on QCC.
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Expressions/SelectExpression.cs]: Should file an issue

File: src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 2]
```diff
@@ -303,6 +303,8 @@ public virtual Expression VisitSelect(SelectExpression selectExpression)
 
             _valueConverterWarningsEnabled = oldValueConverterWarningsEnabled;
 
+            selectExpression.DetachContext();
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Mutating selectExpression during SQL generator is bit risky.

File: test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.cs [hunk 3]
```diff
@@ -118,6 +119,63 @@ private static T Scoper<T>(Func<T> getter)
...
+                        Assert.Throws<InvalidOperationException>(
+                            () => (from c in context1.Customers
+                                   from o in set
+                                   select c).First()).Message);
+                }
+            }
+        }
+
+        [Fact]
+        public virtual void Multiple_context_instances_parameter()
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.cs]: Could add another test where at least one of the queryable is from subquery

Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.cs]: Also a test for DbQuery somewhere?
+ why these tests are not in specs test?

---
In reply to: [180206644](https://github.com/aspnet/EntityFramework/pull/11575#discussion_r180206644) [](ancestors = 180206644)

File: src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 2 above)
Comment (anpete) [reply src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Yes...but less risky than any other place :trollface:

File: test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.cs (hunk 3 above)
Comment (anpete) [reply test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.cs]: Not sure, will try and move.

File: src/EFCore.Relational/Query/Expressions/SelectExpression.cs (hunk 1 above)
Comment (anpete) [reply src/EFCore.Relational/Query/Expressions/SelectExpression.cs]: #11601

**Pull request merged**